### PR TITLE
Add live market liquidity tracking to swap form limits

### PR DESCRIPTION
**Pull request by chibie**
### Description

This PR adds real-time provider liquidity tracking to the swap form, replacing static amount limits with dynamic bounds based on live provider capacity from the aggregator's order book.

**Key Changes:**

1. **New Market Liquidity Module** (`app/lib/marketLiquidity.ts`):
   - `computeLiquidityEnvelope()`: Derives fillable amount ranges from provider offers, accounting for provider floats, exchange rates, and order bands
   - `filterOffersForCorridor()`: Safely filters the aggregator's multi-corridor book to a single corridor
   - `envelopesEqual()`: Prevents unnecessary re-renders by comparing envelope values
   - Message helpers for displaying dynamic limits to users

2. **Markets API Client** (`app/api/aggregator.ts`):
   - `fetchMarkets()`: Fetches the live order book from `/v2/markets` with request deduplication, caching (10s TTL), and envelope-agnostic parsing
   - Handles multiple response shapes gracefully, degrading to empty results on unrecognized formats

3. **Market Liquidity Hook** (`app/hooks/useMarketLiquidity.ts`):
   - Polls the order book every 12 seconds (matching aggregator's cache TTL)
   - Pauses polling when the page is hidden
   - Invalidates stale data when the corridor changes
   - Returns `null` on failures, allowing graceful fallback to static limits

4. **Form Integration** (`app/pages/TransactionForm.tsx`):
   - Integrates `useMarketLiquidity` to fetch live bounds
   - Replaces static limits with dynamic ones when available
   - Displays context-aware messages ("Up to X available right now" vs. static "Maximum is X")
   - Gracefully degrades to pre-existing behavior when the book is unavailable

5. **Button Validation** (`app/hooks/useSwapButton.ts`):
   - Enforces live liquidity bounds on both on-ramp and off-ramp amounts
   - Disables the swap button when the amount exceeds provider capacity or no offers are viable

**Design Rationale:**

- **Null = Unknown**: Empty or unrecognized responses return `null`, not empty envelopes, so the form keeps its static limits rather than locking the user out during outages
- **Envelope, Not Sum**: The range is the envelope across providers (min of all mins, max of all maxes), since a single order is filled by a single provider
- **Inward Rounding**: Bounds round toward the center (ceil min, floor max) so accepted amounts never fail downstream validation
- **Request Deduplication**: Multiple consumers share one request per corridor, with concurrent callers collapsing onto a single promise
- **Visibility-Aware Polling**: Stops polling when the page is hidden to reduce unnecessary requests

**Testing:**

- 418 lines of unit tests covering envelope computation, offer filtering, rounding, real-world book shapes, and message generation
- 141 lines of tests for the markets API client, including envelope parsing tolerance, caching, and request deduplication
- 93 lines of tests for swap button liquidity enforcement on both on-ramp and off-ramp flows
- All tests pass; no existing tests were broken

### References

Addresses the need for dynamic form limits that reflect live provider capacity, preventing users from entering amounts that cannot be filled.

### Testing

- Added comprehensive unit tests in `__tests__/marketLiquidity.test.ts` covering all envelope computation logic, edge cases (rounding, missing balances, dust floats), and real-world book shapes
- Added unit tests in `__tests__/fetchMarkets.test.ts` for API client envelope parsing, caching, and request deduplication
- Added unit tests in `__tests__/useSwapButton.test.tsx` for liquidity band enforcement on swap button state
- All new tests pass; existing tests remain unaffected
- Manual testing: Form limits update in real-time as the order book changes; gracefully falls back to static limits if the markets endpoint is unavailable

- [x] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [x] I have added documentation and tests for new/changed

https://claude.ai/code/session_01WLKfNPF7uimJacjJC9JXfU

## Summary by CodeRabbit

* **New Features**
  * Added live market-liquidity checks to power swap corridor limits and amount validation.
  * Updated the swap button and transaction form to use live liquidity envelopes (min/max, fillable ranges, and “no liquidity” messaging).
  * Added periodic refetching when the page is visible.
* **Bug Fixes**
  * Prevented outdated liquidity responses from affecting current inputs.
  * Improved tolerance of varying aggregator response shapes and error envelopes.
  * Kept the last valid liquidity range during temporary fetch failures.
* **Tests**
  * Expanded coverage for market fetching/parsing, caching/deduping, liquidity math, and fillability validation.

### CODE REVIEW

File: app/pages/TransactionForm.tsx [hunk 1]
```diff
@@ -883,6 +937,11 @@ export const TransactionForm = ({
     rate,
     isSwapped,
     networkName: selectedNetwork.chain.name,
+    liquidity: {
+      min: liquidity?.viable ? liquidity.min : null,
+      max: liquidity?.viable ? liquidity.max : null,
+      noLiquidity,
+    },
   });
```
[the remark this answers is not in the document]
Comment (chibie) [reply app/pages/TransactionForm.tsx]: Fixed in 083a8d5, but flagging one deliberate behavior change for reviewers, since it touches the verification funnel rather than just the bug.

`TransactionForm` now computes the effective bounds once in a `useMemo` and both it and `useSwapButton` read them. Because the CTA reads the merged bounds, it also enforces the **static ceiling it previously ignored**: an unverified user entering an over-cap amount now sees a disabled button alongside the field error, where before the button stayed enabled so they could still reach phone/KYC verification.

That seemed right — an enabled CTA next to a red "Maximum amount is 10,000" is the same class of inconsistency reported here — but it is a funnel change, so it should be a conscious call rather than a side effect. Reverting to liquidity-only bounds on the CTA is a one-line change if you'd rather keep the old behavior.

Also worth noting for anyone tempted by the committable suggestion: it would not compile. `minAmountSentValue`/`maxAmountSentValue` are locals inside `registerFormFields()` within the registration effect, so they aren't in scope at the `useSwapButton` call site — which is why they had to be lifted to component scope instead.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/hooks/useSwapButton.ts [hunk 2]
```diff
@@ -76,11 +102,21 @@ export function useSwapButton({
...
+  const withinBounds =
+    !amountBounds?.noLiquidity &&
+    Number(amountSent) <= amountCeiling &&
+    fitsLiquiditySegment(amountBounds?.segments, Number(amountSent));
   const isAmountValid = isSwapped
     ? !token ||
-      (Number(rate) > 0 && Number(amountSent) >= 0.5 * Number(rate))
-    : Number(amountSent) >= 0.5;
+      (withinBounds && Number(rate) > 0 && Number(amountSent) >= amountFloor)
+    : withinBounds && Number(amountSent) >= amountFloor;
```
[the remark this answers is not in the document]
Comment (chibie) [reply app/hooks/useSwapButton.ts:119]: Confirmed — reproduced with a unit test (insufficient balance + `noLiquidity: true` still returned `isEnabled: true`, CTA text "Fund wallet"). Added the guard right after the connect-wallet states in `app/hooks/useSwapButton.ts`, exactly as suggested, plus a regression test. Pushed in fd4dc0d.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
